bin/textcleaner

```diff
@@ -0,0 +1,619 @@
+# though probably not resolved until the latter
+# so -colorspace HSL/HSB -separate and -colorspace gray became linear
+# but we need to use -set colorspace RGB before using them at appropriate times
+# so that results stay as in original script
+# The following was determined from various version tests using textcleaner
+# with IM 6.7.4.10, 6.7.6.10, 6.7.9.0
+if [ "$im_version" -lt "06070607" -o "$im_version" -gt "06070707" ]; then
+	setcspace="-set colorspace RGB"
+else
+	setcspace=""
+fi
+# no need for setcspace for grayscale or channels after 6.8.5.4
+if [ "$im_version" -gt "06080504" ]; then
+	setcspace=""
+fi
+
+if [ "$invert" != "" ]; then
+	inversion1="-negate"
+else
+	inversion1=""
+fi	
+
+# read the input image into the TMP cached image.
+convert -quiet "$infile" +repage $inversion1 "$tmpA1" ||
+	errMsg "--- FILE $infile NOT READABLE OR HAS ZERO SIZE ---"
+
+# get image size
+ww=`convert $tmpA1 -ping -format "%w" info:`
+hh=`convert $tmpA1 -ping -format "%h" info:`
+
+# get image h/w aspect ratio and determine if portrait=1 (h/w>1) or landscape=0 (h/w<1)
+aspect=`convert xc: -format "%[fx:($hh/$ww)>=1?1:0]" info:`
+ 
+#echo "ww=$ww; hh=$hh; aspect=$aspect"
+
+# set up rotation
+if [ "$layout" = "portrait" -a $aspect -eq 0 -a "$rotate" = "cw" ]; then
+	rotation="-rotate 90"
+elif [ "$layout" = "portrait" -a $aspect -eq 0 -a "$rotate" = "ccw" ]; then
+	rotation="-rotate -90"
+elif [ "$layout" = "landscape" -a $aspect -eq 1 -a "$rotate" = "cw" ]; then
+	rotation="-rotate 90"
+elif [ "$layout" = "landscape" -a $aspect -eq 1 -a "$rotate" = "ccw" ]; then
+	rotation="-rotate -90"
+else
+	rotation=""
+fi
+	
+# set up cropping
+if [ "$cropoff" != "" -a $numcrops -eq 1 ]; then
+	wwc=`convert xc: -format "%[fx:$ww-2*$crop1]" info:`
+	hhc=`convert xc: -format "%[fx:$hh-2*$crop1]" info:`
+	cropping="-crop ${wwc}x${hhc}+$crop1+$crop1 +repage"
+elif [ "$cropoff" != "" -a $numcrops -eq 2 ]; then
+	wwc=`convert xc: -format "%[fx:$ww-2*$crop1]" info:`
+	hhc=`convert xc: -format "%[fx:$hh-2*$crop2]" info:`
+	cropping="-crop ${wwc}x${hhc}+$crop1+$crop2 +repage"
+elif [ "$cropoff" != "" -a $numcrops -eq 4 ]; then
+	wwc=`convert xc: -format "%[fx:$ww-($crop1+$crop3)]" info:`
+	hhc=`convert xc: -format "%[fx:$hh-($crop2+$crop4)]" info:`
+	cropping="-crop ${wwc}x${hhc}+$crop1+$crop2 +repage"
+else
+	cropping=""
+fi
+#echo "cropoff=$cropoff; numcrops=$numcrops; cropping=$cropping"
+
+# test if grayscale
+grayscale=`convert $tmpA1 -format "%[colorspace]" info:`
+typegray=`convert $tmpA1 -format '%r' info: | grep 'Gray'`
+if [ "$gray" = "yes" -o "$grayscale" = "Gray" -o "$typegray" != "" ]; then 
+	makegray="$setcspace -colorspace gray -type grayscale"
+else
+	makegray=""
+fi
+#echo "makegray=$makegray"
+
+# set up enhance
+if [ "$enhance" = "stretch" ]; then
+	enhancing="$setcspace -contrast-stretch 0"
+elif [ "$enhance" = "normalize" ]; then
+	enhancing="$setcspace -normalize"
+else
+	enhancing=""
+fi
+#echo "enhancing=$enhancing"
+
+# setup blurring
```

Comment:
@andjsmit @dlpierce It looks like the "threshold" argument to textcleaner is massaged into a "blurring" argument actually passed to `convert`; I don't see that `textcleaner` actually has a way to set a real `threshold` value for `convert`

Reply:
Okay then. Maybe there's some later optimization that could be done by rolling that option into the textcleaner script.